core/lib/Slots.rei

```diff
@@ -1,8 +1,24 @@
-type t('slot, 'nextSlots);
-let create: unit => t('slot, 'nextSlots);
-let use:
-  (~default: 'slot, t('slot, t('slot2, 'nextSlots))) =>
-  (('slot, ('slot => 'slot) => unit), t('slot2, 'nextSlots));
-let flushPendingUpdates: t('slot, 'nextSlots) => bool;
-
-type empty = t(unit, unit);
+module type Elem = {type t('a);};
+
+module type S = {
+  type elem('a);
+  type opaqueElement =
+    | Any(elem('a)): opaqueElement;
+  type t('slot, 'nextSlots);
+  type empty = t(unit, unit);
+
+  let create: unit => t('slot, 'nextSlots);
+  let use:
+    (~default: elem('slot), t('slot, t('slot2, 'nextSlots))) =>
+    (
+      (elem('slot), (elem('slot) => elem('slot)) => unit),
+      t('slot2, 'nextSlots),
+    );
+
+  let fold:
+    ((opaqueElement, 'a, ~flush: unit => bool) => 'a, 'a, t('b, 'c)) => 'a;
```

Comment:
@wokalski this is awesome -- looks like it could be done!
I'm probably missing something  on this line. What's the reason for the existential type in opaqueElement here? I thought intuitively this would be an `elem('a)` but I guess that would create type difficulties?

Reply:
That's correct. `'a` is an accumulator. `opaqueElement`'s contents can be of any type - we don't know what's the unified type of `slots` at the point where we call `fold`.

Reply:
Oh, then I misinterpreted the `'a`. Makes sense.